### PR TITLE
planner: fix generated column causes a server crash

### LINKED ISSUE #15992: USE_INDEX_MERGE on table with generated column causes a server crash
Consider the following statements:

```sql
CREATE TABLE t0(c0 INT, c1 INT AS (c0));
CREATE INDEX i0 ON t0(c1);
SELECT /*+ USE_INDEX_MERGE(t0, c0_2) */ t0.c0 FROM t0 UNION ALL SELECT 0 FROM t0;
```
Unexpectedly, the `SELECT` causes `tidb-server` to crash with the following stack trace:

```
runtime: goroutine stack exceeds 1000000000-byte limit
fatal error: stack overflow

runtime stack:
runtime.throw(0x29650fa, 0xe)
	/usr/lib/go-1.13/src/runtime/panic.go:774 +0x72
runtime.newstack()
	/usr/lib/go-1.13/src/runtime/stack.go:1046 +0x6e9
runtime.morestack()
	/usr/lib/go-1.13/src/runtime/asm_amd64.s:449 +0x8f

goroutine 341 [running]:
github.com/pingcap/tidb/expression.extractDependentColumns(0xc033bc0000, 0x49243a, 0x52c800, 0x2d61900, 0xc0294a2820, 0x0, 0x0, 0x0)
	/tidb/expression/util.go:96 +0x25c fp=0xc04965a378 sp=0xc04965a370 pc=0x1b0f88c
github.com/pingcap/tidb/expression.extractDependentColumns(0xc033bc0000, 0x492439, 0x52c800, 0x2d61900, 0xc0294a2820, 0x0, 0x0, 0x0)
	/tidb/expression/util.go:101 +0x1c4 fp=0xc04965a3e8 sp=0xc04965a378 pc=0x1b0f7f4
github.com/pingcap/tidb/expression.extractDependentColumns(0xc033bc0000, 0x492438, 0x52c800, 0x2d61900, 0xc0294a2820, 0x0, 0x0, 0x0)
	/tidb/expression/util.go:101 +0x1c4 fp=0xc04965a458 sp=0xc04965a3e8 pc=0x1b0f7f4
github.com/pingcap/tidb/expression.extractDependentColumns(0xc033bc0000, 0x492437, 0x52c800, 0x2d61900, 0xc0294a2820, 0x0, 0x0, 0x0)
[...]
github.com/pingcap/tidb/expression.extractDependentColumns(0xc033bc0000, 0x4923d7, 0x52c800, 0x2d61900, 0xc0294a2820, 0x0, 0x0, 0x0)
	/tidb/expression/util.go:101 +0x1c4 fp=0xc04965cec8 sp=0xc04965ce58 pc=0x1b0f7f4
...additional frames elided...
created by github.com/pingcap/tidb/server.(*Server).Run
	/tidb/server/server.go:333 +0x709
```

Environment:
```
mysql> select tidb_version();
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                          |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v4.0.0-beta.2-160-g82529c886
Git Commit Hash: 82529c886cb1cfff7810476a136acf81284bac6e
Git Branch: master
UTC Build Time: 2020-04-01 11:44:27
GoVersion: go1.13.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve?

Issue Number: close #15992 

Problem Summary:
Consider the following statements:
```sql
CREATE TABLE t0(c0 INT, c1 INT AS (c0));
CREATE INDEX i0 ON t0(c1);
SELECT t0.c0 FROM t0 UNION ALL SELECT 0 FROM t0;
```

The c1.virtualExpr will point to c0.
If we change c0 to c1, then the c1.virtualExpr will be c1. This is wrong and makes the server crash when extractColumn.

### What is changed and how it works?

When setting the `virtualExpr` for the virtual generated column, clone the expression.

### Check List 

Tests 

- Integration test

Side effects

### Release note 
Fix generated column causes a server crash.

### CODE REVIEW

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -2822,7 +2822,7 @@ func (b *PlanBuilder) buildDataSource(ctx context.Context, tn *ast.TableName, as
 				if err != nil {
 					return nil, err
 				}
-				colExpr.VirtualExpr = expr
+				colExpr.VirtualExpr = expr.Clone()
```
Comment (Deardrops) [inline planner/core/logical_plan_builder.go:2825]: Check `expr != nil` before use?

Comment (wjhuang2016) [reply planner/core/logical_plan_builder.go:2825]: If the err is nil, the `expr` shouldn't be nil. This assumption is also followed by other usages.

**Pull request merged**